### PR TITLE
Converge job service

### PR DESCRIPTION
**Pull request by pditommaso**
This PR refactor the `JobService` implementation into specialised version for Docker (dev) and Kubernetes (prod). 

The service should implement the common logic for monitoring job execution e.g. [status](https://github.com/seqeralabs/wave/blob/f1b57a52fdb38632d73a50ec4f25f6ec0f62c049/src/main/groovy/io/seqera/wave/service/job/JobService.groovy#L32-L32) operation. 

The launch of jobs such as transfer, build, etc still required separated strategy because each of them require some ad-hoc logic and config e.g. [TransferStrategy](https://github.com/seqeralabs/wave/blob/f1b57a52fdb38632d73a50ec4f25f6ec0f62c049/src/main/groovy/io/seqera/wave/service/blob/TransferStrategy.groovy#L25-L25), [BuildStrategy](https://github.com/seqeralabs/wave/blob/f1b57a52fdb38632d73a50ec4f25f6ec0f62c049/src/main/groovy/io/seqera/wave/service/builder/BuildStrategy.groovy#L33-L33), etc

### CODE REVIEW

**Comment by munishchouhan** [comment]
Tested successfully with 10 parallel pulls in local

**Comment by pditommaso** [comment]
Has K8s been tested?

**Comment by munishchouhan** [comment]
> Has K8s been tested?

yes, tested in local with k8s
```
% kubectl get jobs -n wave-local
NAME                        COMPLETIONS   DURATION   AGE
transfer-7351616108552790   1/1           35s        12h
transfer-280563bee02982dd   1/1           67s        12h
transfer-24a2bc616981cd5e   1/1           69s        12h
transfer-dba634b91b12fd44   1/1           36s        12h
transfer-2c7cfbf250b59f86   1/1           25s        12h
transfer-e7e6c84249e4e635   1/1           32s        12h
transfer-f2f52311d47cdf81   1/1           13s        12h
transfer-d93dd8a388e7dac7   1/1           17s        12h
transfer-a6307f5ff1c12b87   1/1           15s        12h
transfer-c0bcc063574165e7   1/1           44s        12h
transfer-6cbec5eaebf33db0   1/1           26s        12h
transfer-12b44a73177f4597   1/1           50s        12h
transfer-1bf9c4aac443fdb7   1/1           13s        12h
transfer-5f4c2b654d376ae8   1/1           37s        12h
transfer-6d888b0596a9159c   1/1           25s        12h
transfer-bd5ff874f3b3a861   1/1           35s        12h
transfer-e183eaace5466c3f   1/1           5s         12h
transfer-5b1abf15040f70e6   1/1           44s        12h
transfer-bfcc27f01c6ead02   1/1           34s        12h
transfer-4022f82e0e64bcd9   1/1           5s         12h
transfer-ac03cfe7f1ace92f   1/1           53s        12h
transfer-e4c14925d6f46ab4   1/1           30s        12h
transfer-0293e0a28be15859   1/1           5s         12h
transfer-a24575ed92845b86   1/1           55s        12h
transfer-1c44d6be5e531569   1/1           5s         12h
transfer-9200165b3a6a4a0b   1/1           20s        12h
transfer-d8255ca0124ceda3   1/1           5s         12h
transfer-14548bfc762de864   1/1           6s         12h
transfer-e72da461c5dba288   1/1           5s         12h
transfer-8146cf87b1d2d4a6   1/1           5s         12h
transfer-d3fc3847a36115f6   1/1           8s         12h
transfer-6118759194624f3c   1/1           32s        12h
transfer-8637c64fceac5f18   1/1           24s        12h
transfer-e5a16532f234b8e3   1/1           36s        12h
```

**Comment by pditommaso** [comment]
Now the API is stabilised, we need to add some tests. Especially for handlers logic, both for [scan](https://github.com/seqeralabs/wave/blob/2e5c1af7240673d6550032d96b06a34d19446408/src/main/groovy/io/seqera/wave/service/scan/ContainerScanServiceImpl.groovy#L117-L165) and [container build](https://github.com/seqeralabs/wave/blob/2e5c1af7240673d6550032d96b06a34d19446408/src/main/groovy/io/seqera/wave/service/builder/ContainerBuildServiceImpl.groovy#L364-L395)

**Comment by munishchouhan** [comment]
> Now the API is stabilised, we need to add some tests. Especially for handlers logic, both for [scan](https://github.com/seqeralabs/wave/blob/2e5c1af7240673d6550032d96b06a34d19446408/src/main/groovy/io/seqera/wave/service/scan/ContainerScanServiceImpl.groovy#L117-L165) and [container build](https://github.com/seqeralabs/wave/blob/2e5c1af7240673d6550032d96b06a34d19446408/src/main/groovy/io/seqera/wave/service/builder/ContainerBuildServiceImpl.groovy#L364-L395)

Sure, I will add them

**Comment by pditommaso** [comment]
We are almost there with this PR. Let's add some [tests](https://github.com/seqeralabs/wave/pull/620#issuecomment-2333428842) and make a stress tests in the environment.

**Comment by munishchouhan** [comment]
> We are almost there with this PR. Let's add some [tests](https://github.com/seqeralabs/wave/pull/620#issuecomment-2333428842) and make a stress tests in the environment.

ok i will do that

**Comment by munishchouhan** [comment]
Tested on local with k8s successfully:
```
(base) munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
NAME                       STATUS    COMPLETIONS   DURATION   AGE
build-ab743bde24ef0467-1   Running   0/1           56s        56s
(base) munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get pods -n wave-local
NAME                             READY   STATUS    RESTARTS   AGE
build-ab743bde24ef0467-1-5hlnr   1/1     Running   0          66s
(base) munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
NAME                       STATUS     COMPLETIONS   DURATION   AGE
build-ab743bde24ef0467-1   Complete   1/1           69s        81s
scan-b3beaa6e2ecf          Running    0/1           12s        12s
(base) munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get pods -n wave-local
NAME                             READY   STATUS      RESTARTS   AGE
build-ab743bde24ef0467-1-5hlnr   0/1     Completed   0          84s
scan-b3beaa6e2ecf-8vtk2          1/1     Running     0          15s
(base) munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
NAME                       STATUS     COMPLETIONS   DURATION   AGE
build-ab743bde24ef0467-1   Complete   1/1           69s        117s
scan-b3beaa6e2ecf          Complete   1/1           42s        48s
(base) munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get pods -n wave-local
NAME                             READY   STATUS      RESTARTS   AGE
build-ab743bde24ef0467-1-5hlnr   0/1     Completed   0          119s
scan-b3beaa6e2ecf-8vtk2          0/1     Completed   0          50s
```

**Comment by pditommaso** [comment]
Nice, let's begin a round using K8s cluster + redis. Do you have everything to proceed?

**Comment by munishchouhan** [comment]
> Nice, let's begin a round using K8s cluster + redis. Do you have everything to proceed?

Yes, I am testing on dev

**Comment by munishchouhan** [comment]
testes in dev: found out that backofflimit is not set, so it is defaulting to 6 when build is failing
Added backofflimit and set to 1 for build and scan jobs
```
NAME                       COMPLETIONS   DURATION   AGE
build-xxxxxxxf36769da62-1   1/1           19m        36m
build-xxxxxxx17b194f2bc-1   1/1           13m        36m
build-xxxxxxx382dcd9ecd-1   0/1           36m        36m
build-xxxxxxxea39657729-1   0/1           36m        36m
build-xxxxxxx1f2922cd7d-1   0/1           36m        36m
build-xxxxxxx7a0112017d-6   0/1           36m        36m
build-xxxxxxx7a0112017d-7   0/1           21m        21m
build-xxxxxxx5b8b74140c-1   1/1           16m        21m
build-xxxxxxxadc8c4aae7-1   1/1           14m        21m
build-xxxxxxx9dc4d7364b-1   1/1           10m        21m
build-xxxxxxx329dc7496e-1   1/1           9m56s      21m
build-xxxxxxx48233153d2-1   1/1           9m43s      21m
build-xxxxxxx0b833ddf9e-1   0/1           36m        36m
build-xxxxxxx31255d50d8-1   1/1           12m        36m
build-xxxxxxx54dece5b24-1   1/1           13m        36m
```

**Comment by munishchouhan** [comment]
scan jobs:
```
(base) munish.chouhan@Munishs-MacBook-Pro wave_testing % kubectl get jobs -n wave-build | grep scan
scan-1b8bfbff2731          0/1           1s         1s
scan-6e914c9fc104          1/1           21s        24s
scan-9838d50ef6db          0/1           2s         2s
```

**Comment by munishchouhan** [comment]
Tested with 6 parallel builds in dev successfully 
no related errors in logs

**Review by munishchouhan — APPROVED** [review]

File: src/main/groovy/io/seqera/wave/configuration/BuildConfig.groovy [hunk 1]
```diff
@@ -110,6 +121,9 @@ class BuildConfig {
     @Value('${wave.build.force-compression:false}')
     Boolean forceCompression
 
+    @Value('${wave.build.retryAttempts:1}')
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/configuration/BuildConfig.groovy:124]:
```suggestion
    @Value('${wave.build.retry-attempts:3}')
```

File: src/main/groovy/io/seqera/wave/configuration/ScanConfig.groovy [hunk 2]
```diff
@@ -66,6 +69,9 @@ class ScanConfig {
     @Nullable
     private String severity
 
+    @Value('${wave.scan.retryAttempts:1}')
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/configuration/ScanConfig.groovy:72]:
```suggestion
    @Value('${wave.scan.retry-attempts:1}')
```

Comment (pditommaso) [reply src/main/groovy/io/seqera/wave/configuration/ScanConfig.groovy:72]: Let's change to `retry-attempts` also the one for blob uploads

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/configuration/ScanConfig.groovy:72]: blob has it already

https://github.com/seqeralabs/wave/blob/646b7c54d1318c1ed0d16512783698a01d26e61e/src/main/groovy/io/seqera/wave/configuration/BlobCacheConfig.groovy#L92

should i change it to `retry-attempts`?

Comment (pditommaso) [reply src/main/groovy/io/seqera/wave/configuration/ScanConfig.groovy:72]: yeah, in some cases MN mess up with camelCase notation

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/configuration/ScanConfig.groovy:72]: then should we change this, too?

https://github.com/seqeralabs/wave/blob/646b7c54d1318c1ed0d16512783698a01d26e61e/src/main/groovy/io/seqera/wave/configuration/BuildConfig.groovy#L83

File: src/main/groovy/io/seqera/wave/configuration/BuildConfig.groovy (hunk 1 above)
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/configuration/BuildConfig.groovy]: done

File: src/main/groovy/io/seqera/wave/configuration/ScanConfig.groovy (hunk 2 above)
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/configuration/ScanConfig.groovy]: done

File: src/main/groovy/io/seqera/wave/configuration/ScanConfig.groovy [hunk 3]
```diff
@@ -66,6 +69,9 @@ class ScanConfig {
     @Nullable
     private String severity
 
+    @Value('${wave.scan.retry-attempts:3}')
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/configuration/ScanConfig.groovy:72]:
```suggestion
    @Value('${wave.scan.retry-attempts:1}')
```

1 for scan is enough

**Comment by pditommaso** [comment]
Ok, let's merge! 🚀

**Pull request merged**